SLS-1691: Add Source Code Integration (#202)
DataDog/serverless-plugin-datadog: README.md

To further configure your plugin, use the following custom parameters in your `serverless.yml`:

| Parameter                     | Description                                                                                                                                                                                                                                                                                                                                                                                                                         |
| ----------------------------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |

> | `flushMetricsToLogs`          | Send custom metrics by using logs with the Datadog Forwarder Lambda function (recommended). Defaults to `true`. If you disable this parameter, you must set `DATADOG_API_KEY` in your environment. To use an encrypted key, set `apiKMSKey` or `apiKeySecretArn` in the configuration. `flushMetricsToLogs` is ignored when `addExtension` is true.                                                                                 |

Reviewer: > If you disable this parameter, you must set `DATADOG_API_KEY` in your environment.

Could you also just set the `apiKey` config value?

Author: That's right, just `apiKey` would work.

> | `flushMetricsToLogs`          | Send custom metrics by using logs with the Datadog Forwarder Lambda function (recommended). Defaults to `true`. If you disable this parameter, it's required to set `DATADOG_API_KEY` in your environment, or if encrypted, `apiKMSKey` or `apiKeySecretArn` in the configuration. `flushMetricsToLogs` is ignored when `addExtension` is true.                                                                                          |

Reviewer 2: ```suggestion
| `flushMetricsToLogs`          | Send custom metrics by using logs with the Datadog Forwarder Lambda function (recommended). Defaults to `true`. If you disable this parameter, you must set `DATADOG_API_KEY` in your environment. To use an encrypted key, set `apiKMSKey` or `apiKeySecretArn` in the configuration. `flushMetricsToLogs` is ignored when `addExtension` is true.                                                                                          |
```

> | `site`                        | Set which Datadog site to send data, this is only used when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Possible values are `datadoghq.com`, `datadoghq.eu`, `us3.datadoghq.com`, `us5.datadoghq.com`, and `ddog-gov.com`. The default is `datadoghq.com`.                                                                                                                                                              |

Reviewer 2: ```suggestion
| `site`                        | Set which Datadog site to send data to. This is only used when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Possible values are `datadoghq.com`, `datadoghq.eu`, `us3.datadoghq.com`, `us5.datadoghq.com`, and `ddog-gov.com`. The default is `datadoghq.com`.                                                                                                                                                              |
```

> | `apiKey`                      | Datadog API Key, only needed when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Not required when DATADOG_API_KEY environment variable is set. Defining `apiKey` will add the Datadog API key directly to your Lambda functions as an environment variable. For more information about getting a Datadog API key, see the [API key documentation][3].                                                                     |

Reviewer 2: ```suggestion
| `apiKey`                      | Datadog API key. Only needed when `flushMetricsToLogs` is `false` or `addExtension` is `true`. Not required when the DATADOG_API_KEY environment variable is set. Defining `apiKey` adds the Datadog API key directly to your Lambda functions as an environment variable. For more information about getting a Datadog API key, see the [API key documentation][3].                                                                     |
```

> | `appKey`                      | Datadog App Key, only needed when `monitors` is defined. Not required when DATADOG_APP_KEY environment variable is set.                                                                                                                                                                                                                                                                                                                  |

Reviewer 2: ```suggestion
| `appKey`                      | Datadog app key. Only needed when `monitors` is defined. Not required when the DATADOG_APP_KEY environment variable is set.                                                                                                                                                                                                                                                                                                                  |
```

| `apiKeySecretArn`             | The ARN of the secret storing the Datadog API key in AWS Secrets Manager. Use this parameter if you're storing your secrets in AWS Secrets Manager, and when `flushMetricsToLogs` is `false` or `addExtension` is `true`. If set, remember to add the `secretsmanager:GetSecretValue` permission to the Lambda execution role.                                                                                                      |

> | `apiKMSKey`                   | Datadog API key encrypted using KMS. Use this parameter if you're storing your secrets with KMS, and if `flushMetricsToLogs` is `false` or `addExtension` is `true`. Defining `apiKMSKey` adds the Datadog API Key directly to your Lambda functions as an environment variable.                                                                                                                                                    |

Reviewer: I think the different types of API key settings have gotten a bit confusing. It could be a good idea to clarify the situation a bit, maybe reorganize this part to try to make things more clear.

> | `apiKMSKey`                   | Datadog API Key encrypted using KMS. Use this parameter if you're storing your secrets with KMS, and if `flushMetricsToLogs` is `false` or `addExtension` is `true`. Defining `apiKMSKey` will add the Datadog API Key directly to your Lambda functions as an environment variable.                                                                                                                                                     |

Reviewer 2: ```suggestion
| `apiKMSKey`                   | Datadog API key encrypted using KMS. Use this parameter if you're storing your secrets with KMS, and if `flushMetricsToLogs` is `false` or `addExtension` is `true`. Defining `apiKMSKey` adds the Datadog API Key directly to your Lambda functions as an environment variable.                                                                                                                                                     |
```

> | `captureLambdaPayload`        | (Experimental) This optional setting configures Datadog ingestion for incoming and outgoing AWS Lambda payloads. Function request and response values are added as part of outgoing APM spans sent to Datadog, and require Datadog APM to be configured on your function. Obfuscation of fields in the Lambda payload is available via the `replace_tags` block within `apm_config` settings in [datadog.yaml.][12] Defaults to `false`. |

Reviewer 2: ```suggestion
| `captureLambdaPayload`        | (Experimental) This optional setting configures Datadog ingestion for incoming and outgoing AWS Lambda payloads. Function request and response values are added as part of outgoing APM spans sent to Datadog, and require Datadog APM to be configured on your function. You can obfuscate fields in the Lambda payload by using the `replace_tags` block within `apm_config` settings in [datadog.yaml.][12] Defaults to `false`. |
```

| `addLayers`                   | Whether to install the Datadog Lambda library as a layer. Defaults to `true`. Set to `false` when you plan to package the Datadog Lambda library to your function's deployment package on your own so that you can install a specific version of the Datadog Lambda library ([Python][4] or [Node.js][5]).                                                                                                                               |

> | `addExtension`                | Whether to install the Datadog Lambda Extension as a layer. Defaults to `false`. When enabled, it's required to set the environment variable `DATADOG_API_KEY`, or the configuration values `apiKMSKey` or `apiKeySecretArn`. Learn more about the [Datadog Lambda Extension Layer][8] in the documentation. **Note**: AWS only supports Lambda Extensions for [certain runtimes][9].                                                    |

Reviewer 2: ```suggestion
| `addExtension`                | Whether to install the Datadog Lambda Extension as a layer. Defaults to `false`. When enabled, you must set either the environment variable `DATADOG_API_KEY`, or the configuration values `apiKMSKey` or `apiKeySecretArn`. Learn more about the [Datadog Lambda Extension Layer][8] in the documentation. **Note**: AWS only supports Lambda Extensions for [certain runtimes][9].                                                    |
```

| `enableXrayTracing`           | Set `true` to enable X-Ray tracing on the Lambda functions and API Gateway integrations. Defaults to `false`.                                                                                                                                                                                                                                                                                                                            |
| `enableDDTracing`             | Enable Datadog tracing on the Lambda function. Note: This applies only to integrations using the Datadog Extension. Defaults to `true`.                                                                                                                                                                                                                                                                                                  |

> | `enableDDLogs`                | Enable Datadog log collection for the Lambda function. Note: This setting has no effect on logs sent via the Datadog Forwarder. Defaults to `true`.                                                                                                                                                                                                                                                                                      |

Reviewer 2: ```suggestion
| `enableDDLogs`                | Enable Datadog log collection for the Lambda function. Note: This setting has no effect on logs sent by the Datadog Forwarder. Defaults to `true`.                                                                                                                                                                                                                                                                                      |
```

> | `enableSourceCodeIntegration` | Enable Datadog Source Code integration for the function. Only works for Java and Go.                                                                                                                                                                                                                                                                                                                                                |

Reviewer: 1. Source Code -> source code
2. Can you elaborate on what this actually means/does?
3. For "only works for Java and Go", I would rephrase this slightly to be a bit softer, maybe "This feature is currently only supported for functions written in Java or Go."

Author: Good idea, I actually originally had this but removed it because I felt it was a bit much and couldn't figure out a way to word it. Will do!

| `subscribeToApiGatewayLogs`   | Enable automatic subscription of the Datadog Forwarder to API Gateway log groups. Defaults to `true`.                                                                                                                                                                                                                                                                                                                                    |

> | `subscribeToHttpApiLogs`      | Enable automatic subscription of the Datadog Forwarder to Http-api log groups. Defaults to `true`.                                                                                                                                                                                                                                                                                                                                       |

Reviewer 2: ```suggestion
| `subscribeToHttpApiLogs`      | Enable automatic subscription of the Datadog Forwarder to HTTP API log groups. Defaults to `true`.                                                                                                                                                                                                                                                                                                                                       |
```

> | `subscribeToWebsocketLogs`    | Enable automatic subscription of the Datadog Forwarder to Websocket log groups. Defaults to `true`.                                                                                                                                                                                                                                                                                                                                      |

Reviewer 2: ```suggestion
| `subscribeToWebsocketLogs`    | Enable automatic subscription of the Datadog Forwarder to WebSocket log groups. Defaults to `true`.                                                                                                                                                                                                                                                                                                                                      |
```

> | `forwarderArn`                | Setting this parameter subscribes the given Datadog forwarder to the Lambda functions’ CloudWatch log groups. Required when `enableDDTracing` is set to `true` unless the subscription is otherwise applied. For example, if a Datadog Forwarder subscription is applied via Datadog's AWS Integration, then `forwarderArn` is not required.                                                                                             |

Reviewer 2: ```suggestion
| `forwarderArn`                | Setting this parameter subscribes the given Datadog forwarder to the Lambda functions’ CloudWatch log groups. Required when `enableDDTracing` is set to `true` unless the subscription is otherwise applied. For example, if a Datadog Forwarder subscription is applied using Datadog's AWS Integration, then `forwarderArn` is not required.                                                                                             |
```

> | `integrationTesting`          | Set `true` when running integration tests. This will bypass the validation of the Forwarder ARN and the addition of Datadog Monitor output links. Defaults to `false`.                                                                                                                                                                                                                                                                   |

Reviewer 2: ```suggestion
| `integrationTesting`          | Set `true` when running integration tests. This bypasses the validation of the Forwarder ARN and the addition of Datadog Monitor output links. Defaults to `false`.                                                                                                                                                                                                                                                                   |
```

| `enableTags`                  | When set, automatically tag the Lambda functions with the `service` and `env` tags using the `service` and `stage` values from the serverless application definition. It does NOT override if a `service` or `env` tag already exists. Defaults to `true`.                                                                                                                                                                               |

> | `injectLogContext`            | When set, the lambda layer will automatically patch console.log with Datadog's tracing ids. Defaults to `true`.                                                                                                                                                                                                                                                                                                                          |

Reviewer 2: ```suggestion
| `injectLogContext`            | When set, the Lambda layer automatically patches `console.log` with Datadog's tracing IDs. Defaults to `true`.                                                                                                                                                                                                                                                                                                                          |
```

> | `exclude`                     | When set, this plugin will ignore all specified functions. Use this parameter if you have any functions that should not include Datadog functionality. Defaults to `[]`.                                                                                                                                                                                                                                                                 |

Reviewer 2: ```suggestion
| `exclude`                     | When set, this plugin ignores all specified functions. Use this parameter if you have any functions that should not include Datadog functionality. Defaults to `[]`.                                                                                                                                                                                                                                                                 |
```

> | `enabled`                     | When set to false, the Datadog plugin will stay inactive. Defaults to `true`. You can control this option using an environment variable, e.g. `enabled: ${strToBool(${env:DD_PLUGIN_ENABLED, true})}`, to activate/deactivate the plugin during deployment. Alternatively, you can also use the value passed in through `--stage` to control this option, [see example.](#disable-plugin-for-particular-environment)                     |

Reviewer 2: ```suggestion
| `enabled`                     | When set to `false`, the Datadog plugin stays inactive. Defaults to `true`. You can control this option using an environment variable. For example, use `enabled: ${strToBool(${env:DD_PLUGIN_ENABLED, true})}` to activate/deactivate the plugin during deployment. Alternatively, you can also use the value passed in through `--stage` to control this option—[see example](#disable-plugin-for-particular-environment).                     |
```

> | `monitors`                    | When defined, the Datadog plugin will configure monitors for the deployed function. Requires setting `DATADOG_API_KEY` and `DATADOG_APP_KEY` in your environment. To learn how to define monitors, see [To Enable and Configure a Recommended Serverless Monitor.](#to-enable-and-configure-a-recommended-serverless-monitor)                                                                                                            |

Reviewer 2: ```suggestion
| `monitors`                    | When defined, the Datadog plugin will configure monitors for the deployed function. Requires setting `DATADOG_API_KEY` and `DATADOG_APP_KEY` in your environment. To learn how to define monitors, see [To Enable and Configure a Recommended Serverless Monitor](#to-enable-and-configure-a-recommended-serverless-monitor).                                                                                                            |
```

> | `customHandler`               | When set, the specified handler is set as the handler for all the functions. By deafult, the handler is set to `/opt/nodejs/node_modules/datadog-lambda-js/handler.handler` if `addLayers` is set to `true`, or `node_modules/datadog-lambda-js/dist/handler.handler` if `addLayers` is set to `false`                                                                                                                                   |

Reviewer 2: ```suggestion
| `customHandler`               | When set, the specified handler is set as the handler for all the functions. By default, the handler is set to `/opt/nodejs/node_modules/datadog-lambda-js/handler.handler` if `addLayers` is set to `true`, or `node_modules/datadog-lambda-js/dist/handler.handler` if `addLayers` is set to `false`.                                                                                                                                   |
```
